api/client/service/update.go

```diff
@@ -189,65 +189,116 @@ func updateService(spec *swarm.ServiceSpec, flags *pflag.FlagSet) error {
 	return nil
 }
 
-func updateLabels(flags *pflag.FlagSet, field *map[string]string) {
-	if !flags.Changed(flagLabel) {
-		return
+func anyChanged(flags *pflag.FlagSet, fields ...string) bool {
+	for _, flag := range fields {
+		if flags.Changed(flag) {
+			return true
+		}
 	}
+	return false
+}
 
-	values := flags.Lookup(flagLabel).Value.(*opts.ListOpts).GetAll()
+func updateLabels(flags *pflag.FlagSet, field *map[string]string) {
+	if flags.Changed(flagLabel) {
+		if field == nil {
+			*field = map[string]string{}
+		}
 
-	localLabels := map[string]string{}
-	for key, value := range runconfigopts.ConvertKVStringsToMap(values) {
-		localLabels[key] = value
+		values := flags.Lookup(flagLabel).Value.(*opts.ListOpts).GetAll()
+		for key, value := range runconfigopts.ConvertKVStringsToMap(values) {
+			(*field)[key] = value
+		}
+	}
+
+	if field != nil && flags.Changed(flagLabelRemove) {
+		toRemove, _ := flags.GetStringSlice(flagLabelRemove)
+		for _, label := range toRemove {
+			delete(*field, label)
+		}
 	}
-	*field = localLabels
 }
 
-func anyChanged(flags *pflag.FlagSet, fields ...string) bool {
-	for _, flag := range fields {
-		if flags.Changed(flag) {
-			return true
+func updateEnvironment(flags *pflag.FlagSet, field *[]string) {
+	if flags.Changed(flagEnv) {
+		value := flags.Lookup(flagEnv).Value.(*opts.ListOpts)
+		*field = append(*field, value.GetAll()...)
+	}
+	if flags.Changed(flagEnvRemove) {
+		toRemove, _ := flags.GetStringSlice(flagEnvRemove)
+		for _, envRemove := range toRemove {
+			for i, env := range *field {
+				key := envKey(env)
+				if key == envRemove {
```

Comment:
It might be worth creating a utility function that converts a string slice to `map[string]struct{}`. Then these loop over `toRemove` that appear in each function can be simplified to map lookups.